File: examples/simple_webcam/java/client/SimpleWebcamClient_memory.java
import com.robotraconteur.*;

import java.awt.image.*;
import java.io.*;
import javax.imageio.*;
import javax.swing.*;

import org.opencv.core.*;
import org.opencv.imgcodecs.*;

import experimental.simplewebcam3.*;

// Simple client to read images from a Webcam server
// and display the image.  This example uses the "memory"
// member type

public class SimpleWebcamClient_memory
{

    /**
     * @param args
     */
    public static void main(String[] args)
    {

        String url = "rr+tcp://localhost:22355?service=webcam";
        if (args.length > 0)
        {
            url = args[0];
        }

        // Load the opencv library
        nu.pattern.OpenCV.loadShared();

        // Use ClientNodeSetup to initialize node
        ClientNodeSetup setup = new ClientNodeSetup();

        try
        {
            // Register the service type
            RobotRaconteurNode.s().registerServiceType(new experimental__simplewebcam3Factory());

            // Connect to the service
            Webcam c1 = (Webcam)RobotRaconteurNode.s().connectService(url, null, null, null,
                                                                      "experimental.simplewebcam3.Webcam");

            // Capture an image to the "buffer" and "multidimbuffer"
            WebcamImage_size size = c1.capture_frame_to_buffer();

            // Read the full image from the "buffer" memory
            long l = c1.get_buffer().length();
            UnsignedBytes data = new UnsignedBytes((int)l);
            c1.get_buffer().read(0, data, 0, l);

            // Convert and show the image retrieved from the buffer memory
            Mat frame1 = new Mat(size.height, size.width, CvType.CV_8UC3);
            frame1.put(0, 0, data.value);

            // Show the image
            showImage("buffer", frame1);

            // Read the dimensions of the "multidimbuffer" member
            long[] bufsize = c1.get_multidimbuffer().dimensions();

            // Retrieve the data from the "multidimbuffer"
            MultiDimArray segdata = new MultiDimArray(new int[] {100, 100, 1}, new UnsignedBytes(10000));
            c1.get_multidimbuffer().read(new long[] {10, 10, 0}, segdata, new long[] {0, 0, 0},
                                         new long[] {100, 100, 1});

            // Create a new image to hold the image
            Mat frame2 = new Mat(100, 100, CvType.CV_8U);

            // This will actually give you the transpose of the image because MultiDimArray is stored in column-major
            // order, transpose the image to be the correct orientation
            frame2.put(0, 0, ((UnsignedBytes)segdata.array).value);
            Mat frame3 = frame2.t();

            // Show the image
            showImage("multidimbuffer", frame3);
        }
        finally
        {
            setup.finalize();
        }
    }

    public static void showImage(String title, Mat img)
    {
        MatOfByte matOfByte = new MatOfByte();

        Imgcodecs.imencode(".bmp", img, matOfByte);
        byte[] byteArray = matOfByte.toArray();
        BufferedImage bufImage = null;
        try
        {
            InputStream in = new ByteArrayInputStream(byteArray);
            bufImage = ImageIO.read(in);
            JFrame frame = new JFrame();
            frame.setTitle(title);
            frame.getContentPane().add(new JLabel(new ImageIcon(bufImage)));
            frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
            frame.pack();
            frame.setVisible(true);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }
}

File: examples/simple_webcam/cpp/client/simple_webcam_client_memory.cpp

#include <RobotRaconteur.h>
#include "robotraconteur_generated.h"

#include <opencv2/highgui/highgui.hpp>

// Do not use aliases in header files!
namespace RR = RobotRaconteur;
namespace cam = ::experimental::simplewebcam3;

// Simple client to read images from a Webcam server
// and display the image.  This example uses the "memory"
// member type

// This exmaple expects the simple_webcam_service single camera service to be running

int main(int argc, char* argv[])
{
    std::string url = "rr+tcp://localhost:22355?service=webcam";
    if (argc > 1)
    {
        url = argv[1];
    }

    try
    {
        // Use node setup to help initialize client node
        RR::ClientNodeSetup node_setup(ROBOTRACONTEUR_SERVICE_TYPES);
        // Connect to the service
        cam::WebcamPtr c1 = RR::rr_cast<cam::Webcam>(
            RR::RobotRaconteurNode::s()->ConnectService(url, "", nullptr, NULL, "experimental.simplewebcam3.Webcam"));

        // Capture an image to the "buffer" and "multidimbuffer"
        cam::WebcamImage_sizePtr size = c1->capture_frame_to_buffer();

        // Read the full image from the "buffer" memory
        uint64_t l = c1->get_buffer()->Length();
        RR::RRArrayPtr<uint8_t> data = RR::AllocateRRArray<uint8_t>(l);
        c1->get_buffer()->Read(0, data, 0, l);

        // Convert and show the image retrieved from the buffer memory
        cv::Mat frame1(size->height, size->width, CV_8UC3);
        memcpy(frame1.data, &data->at(0), data->size());
        cv::imshow("buffer", frame1);

        // Read the dimensions of the "multidimbuffer" member
        std::vector<uint64_t> bufsize = c1->get_multidimbuffer()->Dimensions();

        // Retrieve the data from the "multidimbuffer"
        RR::RRMultiDimArrayPtr<uint8_t> segdata = RR::AllocateEmptyRRMultiDimArray<uint8_t>({100, 100, 1});

        c1->get_multidimbuffer()->Read({10, 10, 0}, segdata, {0, 0, 0}, {100, 100, 1});

        // Create a new image and copy data from the WebcamImage
        cv::Mat frame2(100, 100, CV_8U);
        memcpy(frame2.data, &segdata->Array->at(0), segdata->Array->size());

        // Take the transpose because of array ordering difference
        cv::Mat frame3 = frame2.t();

        // Show the Image
        cv::imshow("multidimbuffer", frame3);

        // Wait for user and then close the window
        cv::waitKey();
        cv::destroyAllWindows();

        return 0;
    }
    catch (std::exception& e)
    {
        std::cout << "Error occured in client: " << std::string(e.what()) << std::endl;
        return 1;
    }
}
